bgpd/bgp_debug.c

```diff
@@ -1557,6 +1559,91 @@ DEFUN (no_debug_bgp_update_groups,
 	return CMD_SUCCESS;
 }
 
+DEFUN (debug_bgp_vpn,
+       debug_bgp_vpn_cmd,
+       "debug bgp vpn <leak-from-vrf|leak-to-vrf|rmap-event|label>",
+       DEBUG_STR
+       BGP_STR
+       "VPN routes\n"
+       "leaked from vrf to vpn\n"
+       "leaked to vrf from vpn\n"
+       "route-map updates\n"
+       "labels\n")
+{
+	if (!strcmp(argv[3]->arg, "leak-from-vrf")) {
+		if (vty->node == CONFIG_NODE)
+			DEBUG_ON(vpn, VPN_LEAK_FROM_VRF);
+		else
+			TERM_DEBUG_ON(vpn, VPN_LEAK_FROM_VRF);
+	} else if (!strcmp(argv[3]->arg, "leak-to-vrf")) {
+		if (vty->node == CONFIG_NODE)
+			DEBUG_ON(vpn, VPN_LEAK_TO_VRF);
+		else
+			TERM_DEBUG_ON(vpn, VPN_LEAK_TO_VRF);
+	} else if (!strcmp(argv[3]->arg, "rmap-event")) {
+		if (vty->node == CONFIG_NODE)
+			DEBUG_ON(vpn, VPN_LEAK_RMAP_EVENT);
+		else
+			TERM_DEBUG_ON(vpn, VPN_LEAK_RMAP_EVENT);
+	} else if (!strcmp(argv[3]->arg, "label")) {
+		if (vty->node == CONFIG_NODE)
+			DEBUG_ON(vpn, VPN_LEAK_LABEL);
+		else
+			TERM_DEBUG_ON(vpn, VPN_LEAK_LABEL);
+	} else {
+		vty_out(vty, "%% unknown debug keyword \"%s\"\n", argv[3]->arg);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	if (vty->node != CONFIG_NODE)
+		vty_out(vty, "enabled debug bgp vpn %s\n", argv[3]->arg);
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (no_debug_bgp_vpn,
+       no_debug_bgp_vpn_cmd,
+       "no debug bgp vpn <leak-from-vrf|leak-to-vrf|rmap-event|label>",
```

Comment:
As above. Additionally, you can remove this entire DEFUN by placing `[no]` at the start of the previous command and merely flipping which macro you use (`DEBUG_TERM_ON` / `DEBUG_TERM_OFF` / etc.) based on whether `no` is present.

Reply:
Agreed on [no], but since the debug variables aren't parameterised at run-time, it doesn't seem like a LOC win. It would sure be nice if they were an array indexed by enum so we could do something like
```
#define TYPE_CONF 0x2
#define TYPE_TERM 0x4
enum subsystem {
    as4,
    neighbor,
    packet,
    ...
};
bgp_debug_set(types, subsystem, enable)
```

Reply:
You can delete the entire command definition and body and replace it with a 3-line `if/else` statement, remove the installation call to the CLI graph, and improve future maintainability by not having to update two different functions that are copy-pastes of each other with only minor differences by doing this. But it is just a suggestion and certainly not a blocking issue on the PR.

As for your suggestion I implemented more or less the same in a PR submitted a couple weeks ago (#1808)

Reply:
Is this what you mean? I must be missing something:
```
{
        int idx = 0;
        int enable = 0;
        int notfound = 0;

        if (argv_find(argv, argc, "no", &idx)) {
                if (argv_find(argv, argc, "leak-from-vrf", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_OFF(vpn, VPN_LEAK_FROM_VRF);
                        else
                                TERM_DEBUG_OFF(vpn, VPN_LEAK_FROM_VRF);

                } else if (argv_find(argv, argc, "leak-to-vrf", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_OFF(vpn, VPN_LEAK_TO_VRF);
                        else
                                TERM_DEBUG_OFF(vpn, VPN_LEAK_TO_VRF);
                } else if (argv_find(argv, argc, "rmap-event", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_OFF(vpn, VPN_LEAK_RMAP_EVENT);
                        else
                                TERM_DEBUG_OFF(vpn, VPN_LEAK_RMAP_EVENT);
                } else if (argv_find(argv, argc, "label", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_OFF(vpn, VPN_LEAK_LABEL);
                        else
                                TERM_DEBUG_OFF(vpn, VPN_LEAK_LABEL);
                } else {
                        notfound = 1;
                }
        } else {
                enable = 1;
                if (argv_find(argv, argc, "leak-from-vrf", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_ON(vpn, VPN_LEAK_FROM_VRF);
                        else
                                TERM_DEBUG_ON(vpn, VPN_LEAK_FROM_VRF);
                } else if (argv_find(argv, argc, "leak-to-vrf", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_ON(vpn, VPN_LEAK_TO_VRF);
                        else
                                TERM_DEBUG_ON(vpn, VPN_LEAK_TO_VRF);
                } else if (argv_find(argv, argc, "rmap-event", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_ON(vpn, VPN_LEAK_RMAP_EVENT);
                        else
                                TERM_DEBUG_ON(vpn, VPN_LEAK_RMAP_EVENT);
                } else if (argv_find(argv, argc, "label", &idx)) {
                        if (vty->node == CONFIG_NODE)
                                DEBUG_ON(vpn, VPN_LEAK_LABEL);
                        else
                                TERM_DEBUG_ON(vpn, VPN_LEAK_LABEL);
                } else {
                        notfound = 1;
                }
        }
        if (notfound) {
                vty_out(vty, "%% unknown debug bgp vpn keyword\n");
                return CMD_WARNING_CONFIG_FAILED;
        }

        if (vty->node != CONFIG_NODE)
                vty_out(vty, "%abled debug bgp vpn %s\n",
                        (enable? "en": "dis"), argv[idx]->text);

                 return CMD_SUCCESS;
}
```